File: skills/hypatia/scripts/storage.py
"""Filesystem and JSON primitives shared by the skill scripts."""

import hashlib
import json
import os
from pathlib import Path
from uuid import uuid4


def digest(data: str | bytes) -> str:
    return hashlib.sha256(data.encode("utf-8") if isinstance(data, str) else data).hexdigest()


def json_text(value) -> str:
    return json.dumps(value, ensure_ascii=False, indent=2, allow_nan=False) + "\n"


def read_json(path):
    return json.loads(Path(path).read_bytes())


def inside(root, path) -> Path:
    path = Path(path)
    if path.is_absolute():
        raise ValueError("Artifact paths must be relative")
    base = Path(root).resolve(strict=True)
    result = Path(os.path.abspath(base / path))
    if result == base or not result.is_relative_to(base):
        raise ValueError("Artifact path escapes its root")
    current = base
    for part in result.relative_to(base).parts:
        current /= part
        if current.is_symlink():
            raise ValueError("Symlinks are not accepted for artifacts")
    return result


def atomic_write(path, data: str | bytes) -> None:
    path = Path(path)
    path.parent.mkdir(parents=True, exist_ok=True)
    temporary = path.with_name(f"{path.name}.{uuid4()}.partial")
    with os.fdopen(os.open(temporary, os.O_WRONLY | os.O_CREAT | os.O_EXCL, 0o600), "wb") as stream:
        stream.write(data.encode("utf-8") if isinstance(data, str) else data)
    temporary.replace(path)

File: extensions/hypatia/storage.ts
import { createHash, randomUUID } from "node:crypto";
import { existsSync, lstatSync, mkdirSync, readFileSync, realpathSync, renameSync, writeFileSync } from "node:fs";
import { dirname, isAbsolute, relative, resolve, sep } from "node:path";

export const hash = (data: string | Buffer) => createHash("sha256").update(data).digest("hex");
export const json = (value: unknown) => JSON.stringify(value, null, 2) + "\n";
export const readJson = (path: string): unknown => JSON.parse(readFileSync(path, "utf8"));

export function inside(root: string, path: string): string {
  if (isAbsolute(path)) throw new Error("Artifact paths must be relative");
  const base = realpathSync(root);
  const result = resolve(base, path);
  const rel = relative(base, result);
  if (!rel || rel === ".." || rel.startsWith(`..${sep}`) || isAbsolute(rel))
    throw new Error("Artifact path escapes its root");
  let current = base;
  for (const part of rel.split(sep)) {
    current = resolve(current, part);
    if (existsSync(current) && lstatSync(current).isSymbolicLink())
      throw new Error("Symlinks are not accepted for artifacts");
  }
  return result;
}

export function atomicWrite(path: string, data: string | Buffer) {
  mkdirSync(dirname(path), { recursive: true });
  const temporary = `${path}.${randomUUID()}.partial`;
  writeFileSync(temporary, data, { flag: "wx", mode: 0o600 });
  renameSync(temporary, path);
}
